apollo-router/src/plugins/expose_query_plan.rs

```diff
@@ -0,0 +1,248 @@
+const ENABLED_CONTEXT_KEY: &str = "experimental::expose_query_plan.enabled";
+
+#[derive(Debug, Clone)]
+struct ExposeQueryPlan {
+    enabled: bool,
+}
+
+#[async_trait::async_trait]
+impl Plugin for ExposeQueryPlan {
+    type Config = bool;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        Ok(ExposeQueryPlan {
+            enabled: init.config
+                || std::env::var(ENABLE_EXPOSE_QUERY_PLAN_ENV).as_deref() == Ok("true"),
+        })
+    }
+
+    fn query_planning_service(
+        &self,
+        service: BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError>,
+    ) -> BoxService<QueryPlannerRequest, QueryPlannerResponse, BoxError> {
+        service
+            .map_response(move |res| {
+                if res
+                    .context
+                    .get::<_, bool>(ENABLED_CONTEXT_KEY)
+                    .ok()
+                    .flatten()
+                    .is_some()
+                {
+                    if let QueryPlannerContent::Plan { plan, .. } = &res.content {
+                        res.context
+                            .insert(QUERY_PLAN_CONTEXT_KEY, plan.root.clone())
+                            .unwrap();
+                    }
+                }
+
+                res
+            })
+            .boxed()
+    }
+
+    fn router_service(
+        &self,
+        service: BoxService<RouterRequest, RouterResponse, BoxError>,
+    ) -> BoxService<RouterRequest, RouterResponse, BoxError> {
+        let conf_enabled = self.enabled;
+        service
+            .map_future_with_context(move |req: &RouterRequest| {
+                let is_enabled = conf_enabled && req.originating_request.headers().get(EXPOSE_QUERY_PLAN_HEADER_NAME) == Some(&HeaderValue::from_static("true"));
+                if is_enabled {
+                    req.context.insert(ENABLED_CONTEXT_KEY, true).unwrap();
+                }
+                (req.originating_request.body().query.clone(), is_enabled)
+            }, move |(query, is_enabled): (Option<String>, bool), f| async move {
+                let mut res: Result<RouterResponse, BoxError>  = f.await;
+                res = match res {
+                    Ok(mut res) => {
+                        if is_enabled {
+                            let (parts, stream) = http::Response::from(res.response).into_parts();
+                            let (mut first, rest) = stream.into_future().await;
+
+                            if let Some(first) = &mut first {
+                                if let Some(plan) =
+                                    res.context.get_json_value(QUERY_PLAN_CONTEXT_KEY)
+                                {
+                                    first
+                                        .extensions
+                                        .insert("apolloQueryPlan", json!({ "object": { "kind": "QueryPlan", "node": plan, "text": query } }));
```

Comment:
Whoops! This turns out to have been not quite what was needed from the original requirements which actually ask for `object` and `text` to be at the same level — not nested within each other.

```suggestion
                                        .insert("apolloQueryPlan", json!({ "object": { "kind": "QueryPlan", "node": plan }, "text": query }));
```

Reply:
As an observation, I think this would have been easier to detect in review against the requirements if the JSON blob wasn't wrapped onto one line!  (Something like this, but I wrote this on GitHub, so don't it's not necessarily perfect.)

```suggestion
                                        .insert("apolloQueryPlan", json!({
                                                                           "object": {
                                                                             "kind": "QueryPlan",
                                                                             "node": plan
                                                                           },
                                                                           "text": query
                                                                         }));
```